File: examples/multi_agent_agreement.ts
#!/usr/bin/env npx ts-node
/**
 * Multi-agent agreement with cryptographic proof -- zero setup.
 *
 * Three agents negotiate and co-sign a deployment proposal using JACS.
 * Demonstrates quorum (2-of-3), timeout, independent verification,
 * and a full crypto proof chain.
 *
 * Run:
 *   npx ts-node --compiler-options '{"module":"commonjs","moduleResolution":"node","esModuleInterop":true}' examples/multi_agent_agreement.ts
 */

import { JacsClient } from '../jacsnpm/client';

function main(): void {
  // -- Step 1: Create three ephemeral agents ----------------------------------
  console.log('Step 1 -- Create agents');
  const alice = JacsClient.ephemeral('ring-Ed25519');
  const bob = JacsClient.ephemeral('ring-Ed25519');
  const mediator = JacsClient.ephemeral('ring-Ed25519');

  console.log(`  Alice    : ${alice.agentId}`);
  console.log(`  Bob      : ${bob.agentId}`);
  console.log(`  Mediator : ${mediator.agentId}`);

  // -- Step 2: Alice proposes an agreement ------------------------------------
  console.log('\nStep 2 -- Alice proposes an agreement');
  const proposal = {
    proposal: 'Deploy model v2 to production',
    conditions: ['passes safety audit', 'approved by 2 of 3 signers'],
  };

  const deadline = new Date(Date.now() + 60 * 60 * 1000).toISOString();
  const agentIds = [alice.agentId, bob.agentId, mediator.agentId];

  let agreement = alice.createAgreement(proposal, agentIds, {
    question: 'Do you approve deployment of model v2?',
    context: 'Production rollout pending safety audit sign-off.',
    quorum: 2,
    timeout: deadline,
  });
  console.log(`  Agreement ID : ${agreement.documentId}`);
  console.log(`  Quorum       : 2 of 3`);
  console.log(`  Deadline     : ${deadline}`);

  // -- Step 3: Alice signs ----------------------------------------------------
  console.log('\nStep 3 -- Alice signs');
  agreement = alice.signAgreement(agreement);
  console.log(`  Signed by Alice    (${alice.agentId.substring(0, 12)}...)`);

  // -- Step 4: Bob co-signs ---------------------------------------------------
  console.log('\nStep 4 -- Bob co-signs');
  agreement = bob.signAgreement(agreement);
  console.log(`  Signed by Bob      (${bob.agentId.substring(0, 12)}...)`);

  // -- Step 5: Mediator countersigns ------------------------------------------
  console.log('\nStep 5 -- Mediator countersigns');
  agreement = mediator.signAgreement(agreement);
  console.log(`  Signed by Mediator (${mediator.agentId.substring(0, 12)}...)`);

  // -- Step 6: Inspect agreement status ---------------------------------------
  console.log('\nStep 6 -- Agreement status');
  const doc = JSON.parse(agreement.raw);
  const ag = doc.jacsAgreement;
  const sigCount = ag.signatures?.length ?? 0;
  const quorum = ag.quorum ?? ag.agentIDs.length;
  const complete = sigCount >= quorum;

  console.log(`  Signatures : ${sigCount} of ${ag.agentIDs.length}`);
  console.log(`  Quorum met : ${complete}`);
  for (const sig of ag.signatures ?? []) {
    console.log(
      `    ${sig.agentID.substring(0, 12)}... signed at ${sig.date}  (${sig.signingAlgorithm})`,
    );
  }

  // -- Step 7: Independent self-verification ----------------------------------
  console.log('\nStep 7 -- Independent self-verification');
  for (const [name, client] of [
    ['Alice', alice],
    ['Bob', bob],
    ['Mediator', mediator],
  ] as const) {
    const result = client.verifySelf();
    console.log(`  ${name} verifies self: valid=${result.valid}`);
  }

  // -- Cleanup ----------------------------------------------------------------
  alice.dispose();
  bob.dispose();
  mediator.dispose();
  console.log('\nDone.');
}

main();

File: examples/multi_agent_agreement.py
#!/usr/bin/env python3
"""Multi-agent agreement with cryptographic proof -- zero setup.

Three agents negotiate and co-sign a deployment proposal using JACS.
Demonstrates quorum (2-of-3), timeout, independent verification,
and a full crypto proof chain.

Run:
    python examples/multi_agent_agreement.py
"""

import json
import os
import shutil
import tempfile
from datetime import datetime, timedelta, timezone

from jacs.client import JacsClient

DEMO_PASSWORD = "Demo!Str0ng#Pass42"


def create_agent(name: str, base: str) -> JacsClient:
    """Create a persistent agent in a shared workspace."""
    keys_dir = os.path.join(base, f"{name}_keys")
    os.makedirs(keys_dir, exist_ok=True)
    cfg = os.path.join(base, f"{name}.config.json")

    # Write a config that uses agent-specific keys but shared data dir
    from jacs import SimpleAgent
    agent, info = SimpleAgent.create_agent(
        name=name,
        password=DEMO_PASSWORD,
        algorithm="ring-Ed25519",
        data_directory=os.path.join(base, "shared_data"),
        key_directory=keys_dir,
        config_path=cfg,
    )

    # Load through JacsClient for the high-level API
    client = JacsClient(config_path=cfg)
    print(f"  {name}: {client.agent_id}")
    return client


def main() -> None:
    base = tempfile.mkdtemp(prefix="jacs_demo_")
    os.environ["JACS_PRIVATE_KEY_PASSWORD"] = DEMO_PASSWORD
    print(f"Working directory: {base}\n")

    # -- Step 1: Create three agents -----------------------------------------
    print("Step 1 -- Create agents")
    alice = create_agent("alice", base)
    bob = create_agent("bob", base)
    mediator = create_agent("mediator", base)

    # -- Step 2: Alice proposes an agreement ----------------------------------
    print("\nStep 2 -- Alice proposes an agreement")
    proposal = {
        "proposal": "Deploy model v2 to production",
        "conditions": [
            "passes safety audit",
            "approved by 2 of 3 signers",
        ],
    }
    deadline = (datetime.now(timezone.utc) + timedelta(hours=1)).isoformat()

    agreement = alice.create_agreement(
        document=proposal,
        agent_ids=[alice.agent_id, bob.agent_id, mediator.agent_id],
        question="Do you approve deployment of model v2?",
        context="Production rollout pending safety audit sign-off.",
        quorum=2,
        timeout=deadline,
    )
    print(f"  Agreement ID : {agreement.document_id}")
    print(f"  Quorum       : 2 of 3")
    print(f"  Deadline     : {deadline}")

    # -- Step 3: Alice signs --------------------------------------------------
    print("\nStep 3 -- Alice signs")
    agreement = alice.sign_agreement(agreement)
    print(f"  Signed by Alice ({alice.agent_id[:12]}...)")

    # -- Step 4: Bob co-signs -------------------------------------------------
    print("\nStep 4 -- Bob co-signs")
    agreement = bob.sign_agreement(agreement)
    print(f"  Signed by Bob   ({bob.agent_id[:12]}...)")

    # -- Step 5: Mediator countersigns ----------------------------------------
    print("\nStep 5 -- Mediator countersigns")
    agreement = mediator.sign_agreement(agreement)
    print(f"  Signed by Mediator ({mediator.agent_id[:12]}...)")

    # -- Step 6: Check agreement status ---------------------------------------
    print("\nStep 6 -- Agreement status")
    status = alice.check_agreement(agreement)
    print(f"  Complete : {status.complete}")
    print(f"  Pending  : {status.pending}")
    for s in status.signers:
        label = "signed" if s.signed else "pending"
        print(f"    {s.agent_id[:12]}... {label}" +
              (f"  at {s.signed_at}" if s.signed_at else ""))

    # -- Step 7: Independent verification -------------------------------------
    print("\nStep 7 -- Independent verification")
    for name, client in [("Alice", alice), ("Bob", bob), ("Mediator", mediator)]:
        result = client.verify(agreement.raw_json)
        print(f"  {name} verifies: valid={result.valid}")

    # -- Cleanup --------------------------------------------------------------
    shutil.rmtree(base, ignore_errors=True)
    print(f"\nDone. Temp files cleaned up.")


if __name__ == "__main__":
    main()
